File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays.DotNet/director/Chunks/RaysListChunk.cs
using Microsoft.Extensions.Logging;
using ProjectorRays.Common;
using ProjectorRays.Director;

namespace ProjectorRays.director.Chunks;

public class RaysListChunk : RaysChunk
{
    public uint DataOffset;
    public ushort OffsetTableLen;
    public List<uint> OffsetTable = new();
    public uint ItemsLen;
    public Endianness ItemEndianness;
    public List<BufferView> Items = new();

    public RaysListChunk(RaysDirectorFile? dir, ChunkType type) : base(dir, type) { }

    public override void Read(ReadStream stream)
    {
        ReadHeader(stream);
        ReadOffsetTable(stream);
        ReadItems(stream);
    }

    public virtual void ReadHeader(ReadStream stream)
    {
        DataOffset = stream.ReadUint32();
        OffsetTableLen = stream.ReadUint16();
        ItemsLen = stream.ReadUint32();
        ItemEndianness = stream.Endianness;
    }

    public void ReadOffsetTable(ReadStream stream)
    {
        
        for (int i = 0; i < OffsetTableLen; i++)
            OffsetTable.Add(stream.ReadUint32());

        //Dir?.Logger.LogInformation($"ReadOffsetTable: {string.Join(',', OffsetTable)}");
    }

    public void ReadItems(ReadStream stream)
    {
        for (int i = 0; i < OffsetTableLen; i++)
        {
            uint start = OffsetTable[i];
            uint end = i + 1 < OffsetTableLen ? OffsetTable[i + 1] : ItemsLen;
            int len = (int)(end - start);
            if (len <= 0)
            {
                Dir?.Logger.LogWarning($"Skipping item {i} due to invalid or zero-length span: start={start}, end={end}");
                Items.Add(BufferView.Empty);
                continue;
            }
            Items.Add(stream.ReadByteView(len));
        }
    }










    public string ReadString(int index)
    {
        var view = Items[index];
        return System.Text.Encoding.UTF8.GetString(view.Data, view.Offset, view.Size);
    }

    public string ReadPascalString(int index)
    {
        var view = Items[index];
        var rs = new ReadStream(view);
        var len = rs.ReadUint8();
        return System.Text.Encoding.UTF8.GetString(view.Data, view.Offset + 1, len);
    }

    public ushort ReadUint16(int index)
    {
        var rs = new ReadStream(Items[index], ItemEndianness);
        return rs.ReadUint16();
    }

    public uint ReadUint32(int index)
    {
        var rs = new ReadStream(Items[index], ItemEndianness);
        return rs.ReadUint32();
    }
}

File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays_Python/director/chunks/rays_list_chunk.py
from .rays_chunk import RaysChunk, ChunkType
from ...common.stream import BufferView, ReadStream, Endianness

class RaysListChunk(RaysChunk):
    def __init__(self, dir=None, chunk_type=ChunkType.CastListChunk):
        super().__init__(dir, chunk_type)
        self.data_offset = 0
        self.offset_table_len = 0
        self.offset_table = []
        self.items_len = 0
        self.item_endianness = Endianness.BigEndian
        self.items = []

    def read(self, stream: ReadStream):
        self.read_header(stream)
        self.read_offset_table(stream)
        self.read_items(stream)

    def read_header(self, stream: ReadStream):
        self.data_offset = stream.read_uint32()
        self.offset_table_len = stream.read_uint16()
        self.items_len = stream.read_uint32()
        self.item_endianness = stream.endianness

    def read_offset_table(self, stream: ReadStream):
        for _ in range(self.offset_table_len):
            self.offset_table.append(stream.read_uint32())

    def read_items(self, stream: ReadStream):
        for i in range(self.offset_table_len):
            start = self.offset_table[i]
            end = self.offset_table[i+1] if i + 1 < self.offset_table_len else self.items_len
            length = end - start
            if length <= 0:
                self.items.append(BufferView())
                continue
            self.items.append(BufferView(stream.read_bytes(length), 0, length))

    def write_json(self, writer):
        writer.start_object()
        writer.write_key('offsetTable')
        writer.start_array()
        for o in self.offset_table:
            writer.write_val(o)
        writer.end_array()
        writer.end_object()
